// .ignored/nuxt-og-image/dist/runtime/browserUtil.mjs
import { withBase } from "ufo";
export async function screenshot(browser, options) {
  const page = await browser.newPage({
    colorScheme: options.colorScheme
  });
  await page.setViewportSize({
    width: options.width || 1200,
    height: options.height || 630
  });
  const isHtml = options.html || options.path?.startsWith("html:");
  if (isHtml) {
    const html = options.html || options.path?.substring(5);
    await page.evaluate((html2) => {
      document.open("text/html");
      document.write(html2);
      document.close();
    }, html);
    await page.waitForLoadState("networkidle");
  } else {
    await page.goto(withBase(options.path, options.host), {
      timeout: 1e4,
      waitUntil: "networkidle"
    });
  }
  const screenshotOptions = {
    timeout: 1e4
  };
  if (options.delay)
    await page.waitForTimeout(options.delay);
  if (options.mask) {
    await page.evaluate((mask) => {
      for (const el of document.querySelectorAll(mask))
        el.style.display = "none";
    }, options.mask);
  }
  if (options.selector)
    return await page.locator(options.selector).screenshot(screenshotOptions);
  const screenshot2 = await page.screenshot(screenshotOptions);
  await page.close();
  return screenshot2;
}
