// @nuxthq/studio/dist/theme.cjs
'use strict';

const defu = require('defu');

const supportedFields = {
  /**
   * Raw types
   */
  default: {
    type: "string",
    tags: [
      "@studioInput string"
    ]
  },
  string: {
    type: "string",
    tags: [
      "@studioInput string"
    ]
  },
  number: {
    type: "number",
    tags: [
      "@studioInput number"
    ]
  },
  boolean: {
    type: "boolean",
    tags: [
      "@studioInput boolean"
    ]
  },
  array: {
    type: "array",
    tags: [
      "@studioInput array"
    ]
  },
  "design-token": {
    type: "string",
    tags: [
      "@studioInput design-token"
    ]
  },
  object: {
    type: "object",
    tags: [
      "@studioInput object"
    ]
  },
  file: {
    type: "string",
    tags: [
      "@studioInput file"
    ]
  },
  media: {
    type: "string",
    tags: [
      "@studioInput media"
    ]
  },
  component: {
    type: "string",
    tags: [
      "@studioInput component"
    ]
  },
  icon: {
    type: "string",
    tags: [
      "@studioInput icon"
    ]
  },
  /**
   * Design Tokens
   */
  color: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType color"
    ]
  },
  size: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType size"
    ]
  },
  shadow: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType shadow"
    ]
  },
  opacity: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType opacity"
    ]
  },
  font: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType font"
    ]
  },
  "font-weight": {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType font-weight"
    ]
  },
  "font-size": {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType size"
    ]
  },
  "letter-spacing": {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType size"
    ]
  }
};
function field(schema) {
  if (!schema.type) {
    throw new Error(`Missing type in schema ${JSON.stringify(schema)}`);
  }
  const base = JSON.parse(JSON.stringify(supportedFields[schema.type]));
  const result = defu.defu(base, schema);
  if (!result.tags) {
    result.tags = [];
  }
  if (result.icon) {
    result.tags.push(`@studioIcon ${result.icon}`);
    delete result.icon;
  }
  return {
    $schema: result
  };
}
function group(schema) {
  const result = { ...schema };
  if (result.icon) {
    result.tags = [`@studioIcon ${result.icon}`];
    delete result.icon;
  }
  const fields = {};
  if (result.fields) {
    for (const key of Object.keys(result.fields)) {
      fields[key] = result.fields[key];
    }
    delete result.fields;
  }
  return {
    $schema: result,
    ...fields
  };
}

exports.field = field;
exports.group = group;
