// .ignored/nuxt-og-image/dist/runtime/nitro/renderers/browser.mjs
import { screenshot } from "../../browserUtil.mjs";
import loadBrowserLauncherChunk from "#nuxt-og-image/browser";
const BrowserRenderer = {
  name: "browser",
  createSvg: async function createSvg() {
    throw new Error("Browser provider can't create SVGs.");
  },
  createVNode: async function createVNode() {
    throw new Error("Browser provider can't create VNodes.");
  },
  createPng: async function createPng(options) {
    const launchBrowser = await loadBrowserLauncherChunk();
    if (!launchBrowser) {
      throw new Error("Failed to load browser. Is the `browserProvider` enabled?");
    }
    const browser = await launchBrowser();
    let res = null;
    if (browser) {
      try {
        if (options.html) {
          res = await screenshot(browser, options);
        } else {
          res = await screenshot(browser, {
            ...options,
            host: options.requestOrigin,
            path: `/api/og-image-html?path=${options.path}`
          });
        }
      } finally {
        await browser.close();
      }
    }
    return res;
  }
};
export default BrowserRenderer;
